### PR TITLE
[c10d] Use new style barrier support in c10d/gloo

### PR DESCRIPTION
**Pull request by pietern**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#14294 [c10d] Use new style barrier support in c10d/gloo**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13111509/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14295 [c10d] Remove algorithm caching in ProcessGroupGloo&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13111781/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14296 [c10d][easy] Refer to all work with ProcessGroup prefix&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13128977/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14297 [c10d] Add option structs and timeout field&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13158474/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14298 [c10d] Make ProcessGroup::Work::wait() throw&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13158475/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14386 [c10d] Barrier synchronizes with prior work before completing&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13205022/)

This is the final collective to be ported to the new style where there
is no longer a need to keep a cached algorithm instance around. There
is a follow up change incoming to remove the algorithm caching
functionality in ProcessGroupGloo.

Differential Revision: [D13111509](https://our.internmc.facebook.com/intern/diff/D13111509/)

### CODE REVIEW

**Review by teng-li — COMMENTED** [review]
Can you add a test that is similar to the nccl barrier test to verify the behavior, like launching a couple of collectives, and barrier without waiting on these collective's works. To make sure that the collectives' results are correct?

**Comment by pietern** [comment]
I'll see if I can reuse the test you wrote for the NCCL backend after a rebase.

**Comment by pietern** [comment]
See #14386 for change to barrier behavior to synchronize will all prior work.

File: test/test_c10d.py [hunk 1]
```diff
@@ -970,6 +970,25 @@ def test_timeout_kwarg(self):
         # The barrier will now time output
         self.assertFalse(pg.barrier().wait())
 
+    @unittest.skip("Implementation of this functionality pending; see #14373")
+    def test_barrier_implies_wait(self):
+        store = c10d.FileStore(self.file.name, self.world_size)
+        pg = c10d.ProcessGroupGloo(store, self.rank, self.world_size)
```
Comment (teng-li) [inline test/test_c10d.py]: Just chime in here, I didn't remember the default number of ProcessGroupGloo threads. If it's not 2 or more, can you modify this to ensure that at least two threads are created, now that I saw your pending work changes rely on this.

**Review by teng-li — APPROVED** [review]
Approve to unblock landing, see my comments

**Pull request closed without merging**